bpf/nat: fix current behavior that is silently ignoring errors in a revSNAT context

sahid opened the pull request:
The PR is fixing current behavior of nodeport in context of a revSNAT that is silently ignoring errors to process them to the main path anyway.

- In the first patch we remove usage of DROP_NAT_UNSUPP_PROTO in favor of NAT_PUNT_TO_STACK considering that we want to be conservative in such situation.
- In the second patch we we have determined errors that can happen from a revSNAT
point of view to only accept those that are legitimate.
- The last one clarify comment.

sahid committed:
bpf/nodeport: re-circle back on main path for specified errors

The current behavior is re-circling back to the main path for all
errors returned from the rev-SNAT process, eg. when the revSNAT code
tries to rewrite the packet headers.

In this commit we have determined errors from a rev-SNAT context to
only accept those that are legitimate.

This fix try to be conservative, for that is still accepts
DROP_NAT_NO_MAPPING where in a normal situation should never happen.

Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

brb commented on bpf/lib/nodeport.h:
```diff
@@ -630,6 +630,8 @@ int tail_nodeport_nat_ipv6(struct __ctx_buff *ctx)
 #endif
 	ret = snat_v6_process(ctx, dir, &target);
+	if (ret == DROP_INVALID)
+		goto drop_err;
```
Isn't this handled at the line 647-648?

sahid replied:
No if `dir == NAT_DIR_INGRESS`, right?

brb replied:
Ah, you are right. Have you tried changing the if-statement below from `if (dir == NAT_DIR_INGRESS) {` to `if (dir == NAT_DIR_INGRESS && ret == DROP_NAT_NO_MAPPING) {`?

sahid replied:
Even better. Thanks.

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1687,7 +1687,7 @@ int tail_nodeport_nat_ipv4(struct __ctx_buff *ctx)
 		 * done inside a tail call here.
 		 */
-		if (dir == NAT_DIR_INGRESS) {
+		if (dir == NAT_DIR_INGRESS && ret == DROP_NAT_NO_MAPPING) {
```
Check the code again - I think this change is changing the existing behavior of the SNAT. After this change, for the dir=ingress we will start dropping packets which previously were ignored due to `NAT_PUNT_TO_STACK`.

sahid replied:
I wanted to just use  `if ret == DROP_INVALID` to avoid the tail call to ` CILIUM_CALL_IPV4_FROM_LXC` which is early doing a `revalidate_data_pull()`.

Could you let me know what's the meaning of `NAT_PUNT_TO_STACK`?

brb replied:
`NAT_PUNT_TO_STACK` means that the packet should not be SNAT-ed/rev-SNAT-ed. Basically, it is a packet which proto the SNAT engine cannot handle.

Coming back to your original changes, have you actually measured whether it reduces the verifier complexity?

sahid replied:
Not really, I supposed that as it's calling a `tail call` and then `revalidate_data_pull`.

My understanding is that:  `tail call` are limited in number, add verifier complexity to the prog and bring overhead, and even if `revalidate_data_pull` is called on a new `tail call` there are no reasons to make that extra work. Sounds right?

Do you have an easy way to measure verifier complexity? I was trying with 
```
sudo test/bpf/verifier-test.sh -v -f | grep -e "^Prog" -e processed -e "Proccessed" -e "^=>" -e "over limit"
```

But no luck so far.

sahid replied:
Hello @brb, I'm closing this one since it's based on a change that is not relevant anymore. Thank you Martynas for your help.

brb reviewed (dismissed):
We need to check which `snat_v{4,6}_rev_nat()` return codes are definitely drops (`DROP_INVALID`, `DROP_CSUM_L4`, etc), and which require the packet recircling (`NAT_PUNT_TO_STACK`). Handling just `DROP_INVALID` is a non-complete solution.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -589,6 +589,8 @@ int tail_nodeport_nat_ingress_ipv6(struct __ctx_buff *ctx)
 	ret = snat_v6_rev_nat(ctx, &target);
 	if (IS_ERR(ret)) {
+		if (ret == DROP_INVALID)
+			goto drop_err;
```
AFAIU, and I think it's in line with Martynas's comment, there are several error cases where it's unnecessary to recircle. Shouldn't we instead check the case where we do need to perform the tail call? That would seem more coherent with the code comment just below your changes.

sahid replied:
Hello @brb @pchaigno , How do you feel with the last changes? anychance that you trigger CI to ensure everything is working right?

pchaigno replied:
Code changes look good to me.

pchaigno reviewed (commented):
Thanks for the pull request!

I'm not following *why* you are making those changes. Is it simply to improve performance by avoiding an unnecessary tail call?

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -663,14 +663,16 @@ int tail_nodeport_nat_ingress_ipv6(struct __ctx_buff *ctx)
-		ctx_skip_nodeport_set(ctx);
-		ep_tail_call(ctx, CILIUM_CALL_IPV6_FROM_NETDEV);
-		ret = DROP_MISSED_TAIL_CALL;
+		if (ret == DROP_NAT_NOT_NEEDED || ret == DROP_NAT_NO_MAPPING) {
```
How come this doesn't handle https://github.com/cilium/cilium/blob/07dea5d1da0337ff2e5c27ef993a1e8f5fa41047/bpf/lib/nat.h#L1513-L1517 ?

sahid replied:
For this case we want to drop the packet. I think we are good now. (In opposite to previously where the tail-call was hit for no reasons)

julianwiedmann replied:
> For this case we want to drop the packet. I think we are good now. (In opposite to previously where the tail-call was hit for no reasons)

Why is it ok to drop such packets? Where would they get dropped by the main path otherwise?

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1631,14 +1633,16 @@ int tail_nodeport_nat_ingress_ipv4(struct __ctx_buff *ctx)
-		ctx_skip_nodeport_set(ctx);
-		ep_tail_call(ctx, CILIUM_CALL_IPV4_FROM_NETDEV);
-		ret = DROP_MISSED_TAIL_CALL;
+		if (ret == DROP_NAT_NOT_NEEDED || ret == DROP_NAT_NO_MAPPING) {
```
The NAT code seems to consistently use `NAT_PUNT_TO_STACK` as alias for `DROP_NAT_NOT_NEEDED`. We should be doing the same here (for both IPv4 and IPv6).

sahid replied:
Yes that makes sense, I can perhaps provide a change that completely remove definition of `DROP_NAT_NOT_NEEDED`, sounds good?

julianwiedmann reviewed (changes requested):
Frankly, I'm not convinced that this kind of micro-tuning is justified - when it feels very easy to miss some case where we **do** want to circle back (see below).

The commit message also needs fixing - this code path handles more than just
> For a given reply packet from remote backend

For instance it also does revSNAT for host-originating connections.

zacharysarah commented on bpf/lib/nodeport.h:
```diff
@@ -692,14 +692,22 @@ int tail_nodeport_nat_ingress_ipv6(struct __ctx_buff *ctx)
+		if (ret == NAT_PUNT_TO_STACK ||
+		    /* DROP_NAT_NO_MAPPING and DROP_NAT_UNSUPP_PROTO
+		     * are both unwanted behavior from a revNAT
+		     * context. Let's continue to passing it up to the
```
Use either "in a context" or "from a perspective"
```suggestion
		     * are both unwanted behavior in a revNAT
		     * context. Let's continue to passing it up to the
```

zacharysarah commented on bpf/lib/nodeport.h:
```diff
@@ -692,14 +692,22 @@ int tail_nodeport_nat_ingress_ipv6(struct __ctx_buff *ctx)
+			/* In case of no mapping, recircle back to main path. SNAT
+			 * is very expensive in terms of instructions (since we
+			 * don't have BPF to BPF calls as we use tail calls) and
+			 * complexity, hence this is done inside a tail call here.
```
```suggestion
			/* In case of no mapping, recircle back to main path. SNAT
			 * is very expensive in terms of instructions and
			 * complexity. Consequently, this is done inside a tail call here 
			 (because we don't have BPF to BPF calls).
```

zacharysarah commented on bpf/lib/nodeport.h:
```diff
@@ -1873,14 +1881,22 @@ int tail_nodeport_nat_ingress_ipv4(struct __ctx_buff *ctx)
+		if (ret == NAT_PUNT_TO_STACK ||
+		    /* DROP_NAT_NO_MAPPING and DROP_NAT_UNSUPP_PROTO
+		     * are both unwanted behavior from a revNAT
+		     * context. Let's continue to passing it up to the
```
Use either "in a context" or "from a perspective".
```suggestion
		     * are both unwanted behavior in a revNAT
		     * context. Let's continue to passing it up to the
```

zacharysarah commented on bpf/lib/nodeport.h:
```diff
@@ -1873,14 +1881,22 @@ int tail_nodeport_nat_ingress_ipv4(struct __ctx_buff *ctx)
+			/* In case of no mapping, recircle back to main path. SNAT
+			 * is very expensive in terms of instructions (since we
+			 * don't have BPF to BPF calls as we use tail calls) and
+			 * complexity, hence this is done inside a tail call here.
```
```suggestion
			/* In case of no mapping, recircle back to main path. SNAT
			 * is very expensive in terms of instructions and
			 * complexity. Consequently, this is done inside a tail call here 
			 (because we don't have BPF to BPF calls).
```

zacharysarah reviewed (changes requested):
Some minor tweaks for clarity in comments, otherwise LGTM from a docs perspective. 👍🏻

sahid committed:
nodeport: add clarity in comments when recircle back to main path

This has been suggested by @zacharysarah

Co-authored-by: sarah@corleissen.com
Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

zacharysarah reviewed (approved):
LGTM from a docs perspective 👍🏻

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -693,14 +693,24 @@ int tail_nodeport_nat_ingress_ipv6(struct __ctx_buff *ctx)
+		     * context. Let's continue to passing it up to the
+		     * host, and revisiting this later if needed.
+		     */
+		    ret == DROP_NAT_NO_MAPPING || ret == DROP_NAT_UNSUPP_PROTO) {
```
Instead of the additional check for `DROP_NAT_UNSUPP_PROTO`, how about we change https://github.com/cilium/cilium/blob/0d7406af7bf05e1c0c49c098a3b4e025d927c3e7/bpf/lib/nat.h#L1797 to return `NAT_PUNT_TO_STACK` instead - same as a few lines below for any unknown L4 protocol?

Two reasons why I think that's preferable
- we're not exposing ourselves to some deeper-down code path that returns `DROP_NAT_UNSUPP_PROTO`, and where we *want* to treat it as hard error
- it's slightly less code :)

Could you please spell out clearly in the patch description that current code is silently ignoring actual errors, from eg. when the revSNAT code tries to rewrite the packet headers? We might want to backport this as a fix.
It's also probably best to explain why DROP_NAT_NO_MAPPING can be ignored for now.

sahid replied:
It's all done

sahid committed:
bpf/rev-snat: replace DROP_NAT_UNSUPP_PROTO with NAT_PUNT_TO_STACK

The change is replacing DROP_NAT_UNSUPP_PROTO for rev-SNAT
context. Any returns using DROP_NAT_UNSUPP_PROTO will be replaced with
NAT_PUNT_TO_STACK.

In case that rev-SNAT is not supporting a protocol we want the process
to continue anyway processing the packet to the main path, considering
that an other layer may use it.

NOTE: This does not add any functional changes and keeps the current
behavior that looks to be the less intrusive.

Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -884,7 +884,7 @@ snat_v4_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target, __s8 *
 		if (icmphdr.type != ICMP_ECHO &&
 		    icmphdr.type != ICMP_ECHOREPLY)
-			return DROP_NAT_UNSUPP_PROTO;
+			return NAT_PUNT_TO_STACK;
```
Can we please limit this PR to the revSNAT path? That's where we want to improve the error handling ... otherwise this is just getting too confusing.

Having said that, I believe this change (in the SNAT path) would be great as a follow-on and make other users happy, eg. https://github.com/cilium/cilium/issues/23955 or https://github.com/cilium/cilium/issues/15890

sahid replied:
done

julianwiedmann reviewed (approved):
Thanks!
